Add board

lrharris215 opened the pull request:
The game now prints an empty board to the console when run. It will then print another board with some discs already placed on it, to show off the different colors.

lrharris215 committed:
Create Board class skeleton

Board class has a constructor, and the outline of two methods: printBoard and addDisc.

lrharris215 committed:
Create Disc class

The Disc will represent the ConnectFour pieces. The
constructor takes in a String color and a char symbol.
printDisc will print the symbol to the console.

lrharris215 committed:
Add Presenter class

The Presenter takes care of all print related methods for
the Board and Disc classes. The printBoard and printDisc
methods are static, and format the data printed to the
console. I also removed the redundant methods from Disc
and Board.

lrharris215 committed:
Add NullDisc class

The NullDisc is a Singleton disc that is meant to
represent an empty spot on the Connect Four board. I also
added the Board#setUpGrid method to fill the grid with
NullDiscs at the start of the game. Fixed the formatting
of the Presenter as well.

lrharris215 committed:
Add Discable interface

The Discable interface replaces the inheritance between the NullDisc and Disc.

lrharris215 committed:
Test: Add tests for Disc class

Add tests for the getColor and getSymbol methods of Disc.

lrharris215 committed:
Test: Add tests for NullDisc class

Add tests for getNullDisc, getSymbol, and getColor()

lrharris215 committed:
Test: Add tests for Board Class

Add tests for Board#setUpGrid and Board#findDisc. Also added
Board#findDisc, which returns the disc at a specific
location.

lrharris215 committed:
Add Display

Add a Display class. This will handle all System.out
functions in the game, while making it easier to test. Can
swap in different displays.

lrharris215 committed:
Removed references to testDisc from ConnectFour

testDisc was used to make sure the Board was printing
correctly and is now unneeded.

lrharris215 committed:
Add Displayable Interface, write Presenter Tests

While trying to write tests for my Presenter, I realized
that Display needed an interface so I could make a
TestDisplay. As a result, Display no longer needs to be
initialized with an output. I also called Board#setUpGrid
in the constructor, so I don't need to fill the grid in
separately.

lrharris215 committed:
Add colors

I added some ANSI colors to the presenter, so it can print
the board and discs in color.

lrharris215 committed:
Fix: TestPresenter

Adding colors broke one of the tests, changing the testDisc
color from blue to null fixed it.

kudukacar commented on src/main/java/com/connectfour/Display.java:
```diff
@@ -0,0 +1,24 @@
+    public void println(char message){
+        System.out.println(message);
+    }
+
```
I didn't see any unit tests for Display. In any case, how will you test Display if you don't use dependency injection?  See this example:  https://github.com/kudukacar/server/blob/master/src/main/java/infrastructure/Logger.java and this https://github.com/kudukacar/server/blob/master/src/test/java/infrastructure/LoggerTest.java

lrharris215 replied:
Good point, originally had it like that, then changed it. Will change it back

kudukacar commented on src/test/java/com/connectfour/DiscTest.java:
```diff
@@ -0,0 +1,20 @@
+    void testGetColor() {
+        Disc testDisc = new Disc("blue", 'X');
+        assertEquals("blue", testDisc.getColor());
+    }
```
I don't think you need to test Disc.  It's just a data object.  It does not have any methods (besides getters).

kudukacar commented on src/test/java/com/connectfour/NullDiscTest.java:
```diff
@@ -0,0 +1,26 @@
+        NullDisc testNull = NullDisc.getNullDisc();
+        assertEquals("null", testNull.getColor());
+    }
+}
```
Same comment as above regarding tests for Disc.

kudukacar commented:
You're already a Java expert.  Lots of good stuff here.  Nice use of interfaces!

lrharris215 committed:
Delete: DiscTest and NullDiscTest

Deleted these two test files, because they were
unnecessary. Disc and NullDisc's only methods are getters,
and they don't need to be tested.

lrharris215 committed:
Refactor: Display

Display now takes in a PrintSteam in its constructor. This
will make testing easier.

lrharris215 committed:
Test: Display

The tests for Display take in a TestStream that saves the
input to a state instead of printing to the console.

lrharris215 committed:
Refactor: Added @BeforeEach for tests

Added @BeforeEach for initializing the classes used in all
tests.

kudukacar reviewed (approved):
Looks Good!

lrharris215 committed:
Updated ReadME

Add About section, and instructions on how to run the program and play the game.
